sw/host/rom_ext_image_tools/src/image/image.rs

```diff
@@ -0,0 +1,166 @@
+struct PublicConfig {
+    fields: Vec<Field>,
+}
+
+/// Configuration used for the manifest initialisation.
+#[derive(Deserialize, Debug)]
+struct Field {
+    name: String,
+    value: String,
+    data_type: String,
+}
+
+/// Stripped binary image buffer.
+pub struct RawImage {
+    data: Vec<u8>,
+    input_file: PathBuf,
+}
+
+/// Buffer manipulation API.
+impl RawImage {
+    /// Creates the new image buffer.
+    ///
+    /// The data is read from the requested raw binary file.
+    pub fn new(input_file: &Path) -> Self {
+        let mut f = File::open(input_file)
+            .expect("Failed to open the binary file!");
+        let mut data = Vec::new();
+
+        // Read the whole file
+        f.read_to_end(&mut data)
+            .expect("Failed to read the binary file!");
+
+        RawImage {
+            data: data,
+            input_file: input_file.to_path_buf(),
+        }
+    }
+
+    /// Updates the buffer with the supplied config data.
+    ///
+    /// The config data is parsed from the requested config file, and the
+    /// corresponding the manifest portion of the image buffer is updated.
+    pub fn update(&mut self, input_file: &Path) {
+        let mut f = File::open(input_file)
+            .expect("Failed to open the config file!");
+
+        let mut data = String::new();
+
+        // Read the whole configuration file.
+        f.read_to_string(&mut data)
+            .expect("Failed to read the config file!");
+
+        let data: Value = serde_hjson::from_str(&data)
+            .expect("Failed to parse the hjson config file!");
+
+        let deserialized: PublicConfig = serde_hjson::from_value(data)
+            .expect("Failed to deserialize hjson config data!");
+
+        self.splice_config(&deserialized);
+    }
+
+    /// Writes the image buffer contents into a file.
+    ///
+    /// Places the new file alongside the original, with a "new_" prefix.
+    pub fn write_file(&self) {
+        let file_name = self.input_file.file_name()
+            .expect("Failed to get file stem!");
+
+        let mut new_file_name = OsString::from("new_");
+        new_file_name.push(file_name);
+
+        let output_file = self.input_file.with_file_name(new_file_name);
+
+        let mut f = File::create(output_file).
+            expect("Failed to created the new binary file!");
+
+        f.write_all(&self.data)
+            .expect("Failed to write the new binary file!");
+    }
+
+    /// Updates the manifest portion of the image buffer.
+    ///
+    /// This function loops through all the fields supplied through the config,
+    /// finds a matching field data in the manifest, and parses the fields.
+    fn splice_config(&mut self, config: &PublicConfig) {
+        // Loop through the config fields.
+        for field in &config.fields {
+            let field_name = field.name.as_str();
+            let manifest_data =
+                match manifest::get_field_from_name(field_name) {
```

Comment:
Writing `get_field_from_name(&field.name)` will Deref-coerce the &String to &str, so you can avoid the as_str() call.